### PR TITLE
feat(big-query): add bigquery client and getEmailsData method

### PR DESCRIPTION
**Pull request by kapoc3**
Add client for big query with getEmailsData method for bigquery authorization page fetch data

### CODE REVIEW

File: .env.development [hunk 1]
```diff
@@ -6,6 +6,7 @@ REACT_APP_USE_DOPPLER_LEGACY_LOGIN=false
 REACT_APP_USE_DOPPLER_LEGACY_SIGNUP=false
 REACT_APP_USE_DOPPLER_LEGACY_FORGOTPASSWORD=false
 REACT_APP_SHOPIFY_URL= https://sfy.fromdoppler.com:4433
+REACT_APP_BIGQUERY_URL= https://bqy.fromdoppler.com:4433
```
Comment (andresmoschini) [inline .env.development:9]:
```suggestion
REACT_APP_BIGQUERY_URL= https://apisint.fromdoppler.net/bigquery
```

File: .env.int [hunk 2]
```diff
@@ -6,6 +6,7 @@ REACT_APP_USE_DOPPLER_LEGACY_LOGIN=false
 REACT_APP_USE_DOPPLER_LEGACY_SIGNUP=false
 REACT_APP_USE_DOPPLER_LEGACY_FORGOTPASSWORD=false
 REACT_APP_SHOPIFY_URL= https://sfy.fromdoppler.com
+REACT_APP_BIGQUERY_URL= https://bqy.fromdoppler.com:4433
```
Comment (andresmoschini) [inline .env.int:9]:
```suggestion
REACT_APP_BIGQUERY_URL= https://apisint.fromdoppler.net/big-query
```

File: .env.production [hunk 3]
```diff
@@ -6,6 +6,7 @@ REACT_APP_USE_DOPPLER_LEGACY_LOGIN=false
 REACT_APP_USE_DOPPLER_LEGACY_SIGNUP=false
 REACT_APP_USE_DOPPLER_LEGACY_FORGOTPASSWORD=false
 REACT_APP_SHOPIFY_URL= https://sfy.fromdoppler.com
+REACT_APP_BIGQUERY_URL= https://bqy.fromdoppler.com:4433
```
Comment (andresmoschini) [inline .env.production:9]:
```suggestion
REACT_APP_BIGQUERY_URL= https://apis.fromdoppler.com/big-query
```

File: .env.qa [hunk 4]
```diff
@@ -6,6 +6,7 @@ REACT_APP_USE_DOPPLER_LEGACY_LOGIN=false
 REACT_APP_USE_DOPPLER_LEGACY_SIGNUP=false
 REACT_APP_USE_DOPPLER_LEGACY_FORGOTPASSWORD=false
 REACT_APP_SHOPIFY_URL= https://sfy.fromdoppler.com
+REACT_APP_BIGQUERY_URL= https://bqy.fromdoppler.com:4433
```
Comment (andresmoschini) [inline .env.qa:9]:
```suggestion
REACT_APP_BIGQUERY_URL= https://apisqa.fromdoppler.net/big-query
```

File: src/components/Integrations/BigQuery/ControlPanel/AuthorizationForm.js [hunk 5]
```diff
@@ -7,7 +7,7 @@ export const AuthorizationForm = ({ emails }) => {
   const [errorMessage, setErrorMessage] = useState('');
   const intl = useIntl();
   const _ = (id, values) => intl.formatMessage({ id: id }, values);
-  const initialValues = [...emails, ''];
+  const initialValues = [...emails, ''] || [''];
```
Comment (andresmoschini) [inline src/components/Integrations/BigQuery/ControlPanel/AuthorizationForm.js:10]: Esto está raro, en que casos `[...emails, '']` podría ser _falsy_?

Comment (kapoc3) [reply src/components/Integrations/BigQuery/ControlPanel/AuthorizationForm.js:10]: si, esto era parte de una prueba, se elimina el opcional vacio, por que siempre se carga el valor desde el fetch inicial en el caso de no existir datos ya se envia el array vacio

File: .env.development (hunk 1 above)
Comment (kapoc3) [reply .env.development:9]: change accepted and applied

File: .env.int (hunk 2 above)
Comment (kapoc3) [reply .env.int:9]: change accepted and applied

File: .env.production (hunk 3 above)
Comment (kapoc3) [reply .env.production:9]: change accepted and applied

File: .env.qa (hunk 4 above)
Comment (kapoc3) [reply .env.qa:9]: change accepted and applied

File: src/services/generic-client.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
+import { AxiosInstance, AxiosStatic } from 'axios';
+import { AppSession } from './app-session';
+import { RefObject } from 'react';
+
+interface ApiConnectionData {
+  jwtToken: string;
+}
+
+export class GenericClient {
```
Comment (andresmoschini) [inline src/services/generic-client.ts:9]: Good idea @kapoc3, it is nice! 

In my opinion, I think that this kind of code refactoring should be in a different PR, but I think that it is acceptable.

@cbernat @jcamposmk what do you think?

BTW, what do you think about naming it _HttpGenericClient_ or _HttpBaseClient_?

```suggestion
export class HttpBaseClient {
```

File: src/services/big-query-client.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+  private mapEmails(response: any): EmailList {
+    return {
+      emails: response.emails,
+    };
+  }
+  public async getEmailsData(idUser: String): Promise<ResultWithoutExpectedErrors<EmailList>> {
+    try {
+      const { jwtToken } = super.getApiConnectionData();
+      const response = await this.axios.request({
```
Comment (andresmoschini) [inline src/services/big-query-client.ts:24]: Why do you use `super` for using `getApiConnectionData` and `this` for using `axios`?

Would it work?

```suggestion
      const { jwtToken } = this.getApiConnectionData();
      const response = await this.axios.request({
```

File: src/services/big-query-client.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        method: 'GET',
+        url: `/big-query/emails`,
```
Comment (andresmoschini) [inline src/services/big-query-client.ts:26]: The URL should include the accountname (email) because it helps a lot in the security implementation of our APIs:

```suggestion
      const { email, jwtToken } = super.getApiConnectionData();
      const response = await this.axios.request({
        method: 'GET',
        url: `/big-query/{email}/emails`,
```

Comment (andresmoschini) [inline src/services/big-query-client.ts:26]: What do you think about a more descriptive name for this URL?

For example: 

```suggestion
        url: `/big-query/allowed-emails`,
```

File: src/services/big-query-client.ts [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+        params: { idUser: idUser },
+        headers: { Authorization: `token ${jwtToken}` },
+      });
+
+      return this.responseCache.getCachedOrMap(this.getEmailsData, response, (r) => {
+        return r.data && r.data.length
+          ? { success: true, value: r.data.map(this.mapEmails) }
+          : { success: true, value: [''] };
```
Comment (andresmoschini) [inline src/services/big-query-client.ts:34]: I am not sure if using `responseCache` is useful here.

The idea behind this is to use the same object's reference when the response is the same.

It will work, but I think that it is not necessary here. I have no a strong opinion about it.

File: src/services/generic-client.ts (hunk 6 above)
Comment (andresmoschini) [reply src/services/generic-client.ts:9]: On the other hand, in general, I prefer composition over inheritance, is there a way of doing this using composition?

File: src/services/big-query-client.ts (hunk 7 above)
Comment (kapoc3) [reply src/services/big-query-client.ts:24]: in this case is because the method implementation exist on the father class, with this it works but it is not readable that the method is implemented in the parent...

Comment (andresmoschini) [reply src/services/big-query-client.ts:24]: But `this` allows to override it, and it is desirable, right?

File: src/services/big-query-client.ts (hunk 8 above)
Comment (kapoc3) [reply src/services/big-query-client.ts:26]: all apis methods must always have that url param?

Comment (kapoc3) [reply src/services/big-query-client.ts:26]: change accepted and applied

Comment (andresmoschini) [reply src/services/big-query-client.ts:26]: if the result is related to data of an specific account, yes.

We can talk about the reason.

File: src/services/big-query-client.ts (hunk 9 above)
Comment (kapoc3) [reply src/services/big-query-client.ts:34]: the cache was removed

File: src/services/generic-client.ts (hunk 6 above)
Comment (kapoc3) [reply src/services/generic-client.ts:9]: i can change inheritance for composition way, do you prefeer composition?

Comment (andresmoschini) [reply src/services/generic-client.ts:9]: In general, I prefer composition, but in this case, I am not sure if composition applies. So, if you prefer inheritance, I am fine.

File: src/App.js [hunk 10]
```diff
@@ -28,7 +28,7 @@ import InvoicesList from './components/InvoicesList/InvoicesList';
 import ExclusiveForm from './components/DopplerPlus/ExclusiveForm';
 import { ContactPolicy } from './components/ContactPolicy/ContactPolicy';
 import { Promotional } from './components/Integrations/BigQuery/Promotional/Promotional';
-import { AuthorizationPage } from './components/Integrations/BigQuery/ControlPanel/AuthorizationPage';
+import AuthorizationPage from './components/Integrations/BigQuery/ControlPanel/AuthorizationPage';
```
Comment (andresmoschini) [inline src/App.js:31]: I think that we are trying to avoid usage of _default exports_, see the rationale here: https://humanwhocodes.com/blog/2019/01/stop-using-default-exports-javascript-module/

**Comment by andresmoschini** [comment]
In my opinion, it is fine, but I would like to see @cbernat or @jcamposmk review here.

File: src/App.js (hunk 10 above)
Comment (kapoc3) [reply src/App.js:31]: when using the dependency injector is the way to export the class in the solution, i can make this change, but it would not be consistent with the rest of the components of the solution, do you agree with this change

Posible change 
App.js
imagen

new way to export 

imagen


do you agree with this change ?

File: src/services/big-query-client.ts (hunk 7 above)
Comment (kapoc3) [reply src/services/big-query-client.ts]: suggested accepted and applied

File: src/services/big-query-client.ts (hunk 8 above)
Comment (kapoc3) [reply src/services/big-query-client.ts]: the email was incorporated in the query string

File: src/App.js (hunk 10 above)
Comment (andresmoschini) [reply src/App.js:31]: Oh, I understand. @cbernat @jcamposmk what is your opinion?

File: src/services/big-query-client.ts (hunk 8 above)
Comment (andresmoschini) [reply src/services/big-query-client.ts]: Thanks!

Just a minor correction, you have incorporated the email in the path:

```
          userinfo       host      port
          ┌──┴───┐ ┌──────┴──────┐ ┌┴┐
  https://john.doe@www.example.com:123/forum/questions/?tag=networking&order=newest#top
  └─┬─┘   └───────────┬──────────────┘└───────┬───────┘ └───────────┬─────────────┘ └┬┘
  scheme          authority                  path                 query           fragment
```

File: src/App.js (hunk 10 above)
Comment (cbernat) [reply src/App.js:31]: Why not do this?
```
export const AuthorizationPage = InjectAppServices ( ( {dependencies: bigQueryClient .... 
```

**Comment by cbernat** [comment]
👍

File: src/components/Integrations/BigQuery/ControlPanel/AuthorizationPage.js [hunk 11]
```diff
@@ -1,21 +1,24 @@
 import React, { useState, useEffect } from 'react';
 import { Loading } from '../../../Loading/Loading';
-import useTimeout from '../../../../hooks/useTimeout';
 import { AuthorizationForm } from './AuthorizationForm';
+import { InjectAppServices } from '../../../../services/pure-di';
 
-export const AuthorizationPage = () => {
-  const [data, setData] = useState([]);
+const AuthorizationPage = ({ dependencies: { bigQueryClient } }) => {
```
[the remark this answers is not in the document]
Comment (kapoc3) [reply src/components/Integrations/BigQuery/ControlPanel/AuthorizationPage.js]: Solved !!

**Comment by jcamposmk** [comment]
In my opinion it's fine, the tests can be done in a better way without using timeout. Tomorrow I meet with Cristian
